onnxruntime/core/graph/function_utils.cc

```diff
@@ -0,0 +1,672 @@
+    const std::unordered_map<std::string, int>& domain_version_map,
+    const SchemaRegistryManager& schema_registry,
+    const logging::Logger& logger,
+    bool allow_released_opsets_only) {
+  std::string func_identifier = function_utils::GetFunctionIdentifier(function_domain, function_name);
+  auto iter = model_local_functions.find(func_identifier);
+  if (iter == model_local_functions.end()) {
+    ORT_THROW("The given function name: ", function_name, ", domain: ", function_domain, " is not found in model local functions");
+  }
+
+  auto* onnx_func_proto = iter->second;
+  ORT_ENFORCE(onnx_func_proto);
+  // generate the schema for this function template
+  // For schema defined functions get the version from the node in parent graph.
+  // For the functions which do not have schema defined (model local functions)
+  // get the since version from the version in opset imports using the domain.
+  auto since_version = GetVersionForDomain(function_domain, domain_version_map);
+  auto op_schema = std::make_unique<ONNX_NAMESPACE::OpSchema>();
+  op_schema->SetName(function_name);
+  op_schema->SetDomain(function_domain);
+  op_schema->SetDoc(onnx_func_proto->doc_string());
+  op_schema->SinceVersion(static_cast<ONNX_NAMESPACE::OperatorSetVersion>(since_version));
+  std::unordered_map<std::string, int> input_name_idx_map;
+  std::unordered_map<std::string, int> output_name_idx_map;
+
+  for (int i = 0; i < onnx_func_proto->input_size(); ++i) {
+    input_name_idx_map[onnx_func_proto->input().Get(i)] = i;
+  }
+  for (int i = 0; i < onnx_func_proto->output_size(); ++i) {
+    output_name_idx_map[onnx_func_proto->output().Get(i)] = i;
+  }
+
+  // Infer a op_schema for stand-alone functions.
+  IOTypeConstraintHelper(*onnx_func_proto, op_schema, input_name_idx_map, output_name_idx_map);
+  auto allow_official_onnx_release_only_final =
+      allow_released_opsets_only && model_load_utils::IsAllowReleasedONNXOpsetsOnlySet();
+
+  const auto onnx_released_versions =
+      schema_registry.GetLastReleasedOpsetVersions(false);
+
+  std::unordered_map<std::string, int> func_domain_to_version;
+  for (auto& opSet : onnx_func_proto->opset_import()) {
+    const auto& domain = opSet.domain();
+    const auto version = gsl::narrow_cast<int>(opSet.version());
+
+    model_load_utils::ValidateOpsetForDomain(onnx_released_versions, logger,
+                                             allow_official_onnx_release_only_final, domain, version);
+
+    // We need to overwrite the domain here with ("") or else the loop below will try to find ("")
+    // in the map and if not found (when domain == kOnnxDomainAlias), adds an entry for ("", 11).
+    // This effectively ignores the opset version specified by the model for the onnx domain.
+    if (domain == kOnnxDomainAlias) {
+      func_domain_to_version[kOnnxDomain] = version;
+    } else {
+      func_domain_to_version[domain] = version;
+    }
+  }
```

Comment:
I think you can extract this piece of code to a separated function called "GetDomainToVersion".

Reply:
sure